lib/whitespace.coffee

```diff
@@ -77,14 +77,42 @@ class Whitespace
 
   convertTabsToSpaces: (editor) ->
     buffer = editor.getBuffer()
-    spacesText = new Array(editor.getTabLength() + 1).join(' ')
+
+    tabLength = editor.getTabLength()
 
     buffer.transact ->
-      buffer.scan /\t/g, ({replace}) -> replace(spacesText)
+      buffer.scan /^.*\t.*$/g, ({matchText, replace}) ->
+        while match = /^([^\t]*)\t(.*)$/.exec(matchText)
+          newTabLength = tabLength-(match[1].length %% tabLength)
+          matchText = match[1] + (new Array(newTabLength+1).join(" ")) + match[2]
+
+        replace(matchText)
 
   convertSpacesToTabs: (editor) ->
     buffer = editor.getBuffer()
-    spacesText = new Array(editor.getTabLength() + 1).join(' ')
+
+    tabLength = editor.getTabLength()
 
     buffer.transact ->
-      buffer.scan new RegExp(spacesText, 'g'), ({replace}) -> replace('\t')
+      buffer.scan /^.*[ ].*$/g, ({match, replace}) ->
```

Comment:
Similar question here... why not search for chunks of contiguous leading whitespace and use the range to break them up in a tab-stop aware way? Then you don't have to process a bunch of unrelated characters. I don't think we should be inserting tabs for anything but leading whitespace.